Add button to save current view configuration as default in view mode

MichaelBuessemeyer opened the pull request:
### URL of deployed dev instance (used for testing):
- https://updatedefaultlayerconfiginviewmode.webknossos.xyz

### Steps to test:
- Open a dataset in view mode (as admin or dataset manager)
- Change the view mode configuration in a recognizable way (e.g. set histogram min = 100, max = 200)
- Use the new button at the bottom of the layer configuration tab to save the new configuration
- A modal should open up for confirmation. If you cancel the action, no update should be performed.
- If you confirm the modal, a toast should indicate that the updated succeeded
- Use the crog-icon next to the dataset name at the top of the info tab to open the dataset settings
- Go to the view configurations and confirm that the update worked successfully.


## TODOs:
- [x] Test whether this pr works well together with the new layer ordering pr during cover mode for color layers. #7188


### Issues:
- fixes #5183

------
(Please delete unneeded items, merge only when none are left open)
- [x] Updated [changelog](../blob/master/CHANGELOG.unreleased.md#unreleased)
- [x] Updated [documentation](../blob/master/docs) if applicable

MichaelBuessemeyer commented:
@normanrz could you please take a better example image that has the same style as the other images from the docs once this pr is merged?

daniel-wer commented on frontend/javascripts/oxalis/view/left-border-tabs/layer_settings_tab.tsx:
```diff
@@ -1103,6 +1141,17 @@ class DatasetSettings extends React.PureComponent<DatasetSettingsProps, State> {
           </Row>
         ) : null}
 
+        {this.props.controlMode === ControlModeEnum.VIEW && this.props.isAdminOrManager ? (
```
Is there a reason why this is only allowed from view mode? I cannot read from the issue description that this was a requirement and was wondering why you introduced this restriction.
I guess annotations would complicate this functionality, because volume annotation layers would need to be filtered out and/or replaced by their fallback layers :thinking: 
@normanrz Do you think it's sufficient to enable this for viewed datasets, only?

MichaelBuessemeyer replied:
> I guess annotations would complicate this functionality, because volume annotation layers would need to be filtered out and/or replaced by their fallback layers 

Excactly that was my thought :+1: (Additionally norman mentioned to me that view mode only should be fine)

daniel-wer commented on frontend/javascripts/oxalis/view/left-border-tabs/layer_settings_tab.tsx:
```diff
@@ -1049,6 +1052,41 @@ class DatasetSettings extends React.PureComponent<DatasetSettingsProps, State> {
+          Toast.error(
+            "Failed to save the current view configuration as default. Please look at the console for more details.",
+          );
+          console.error(error);
```
Please add airbrake logging as well.

daniel-wer reviewed (commented):
Nice functionality :+1: Please see my suggestions and remarks.

I'll report back after testing :monkey:

daniel-wer commented:
Regarding my comment https://github.com/scalableminds/webknossos/pull/7205#discussion_r1265355156

I tested which settings are saved and restored when using the new functionality and it seems to be all-of-them (all dataset-related settings). This is in contrast to the existing configuration UI where only the values you mentioned can be modified. We need to decide whether (a) it's fine that the new functionality is more powerful, (b) the existing UI should be extended to have the same capabilities, or (c) the new functionality should be restricted to only persist the values that can be modified in the existing UI.

I made a screenshot of the settings before. Then, I modified all settings, saved using the new functionality and opened the dataset with a new user:

Before             |  After
:-------------------------:|:-------------------------:
![settings_before](https://github.com/scalableminds/webknossos/assets/1702075/7792b393-1d42-4725-8b0a-631bc45c73e5)  |  ![settings_after](https://github.com/scalableminds/webknossos/assets/1702075/90a2310b-82a2-4950-8316-4eb3255c990e)
![settings_before_2](https://github.com/scalableminds/webknossos/assets/1702075/30a42f4b-5f3e-4e37-a22b-a62874450c8e)  |  ![settings_after_2](https://github.com/scalableminds/webknossos/assets/1702075/73cf9c04-6c57-4abf-bce9-d9ed2a40645d)

Apart from that inconsistency, the functionality works very well :)

MichaelBuessemeyer commented on frontend/javascripts/oxalis/view/left-border-tabs/layer_settings_tab.tsx:
```diff
@@ -1141,6 +1190,7 @@ const mapStateToProps = (state: OxalisState) => ({
   task: state.task,
   controlMode: state.temporaryConfiguration.controlMode,
   isArbitraryMode: Constants.MODES_ARBITRARY.includes(state.temporaryConfiguration.viewMode),
+  isAdminOrManager: state.activeUser != null ? Utils.isUserAdminOrManager(state.activeUser) : false,
```
Oh yeah, sorry I thought that the "Manager" in `isUserAdminOrManager` stands for "DatasetManager". Thanks for pointing out

MichaelBuessemeyer commented on frontend/javascripts/oxalis/view/left-border-tabs/layer_settings_tab.tsx:
```diff
@@ -1049,6 +1052,41 @@ class DatasetSettings extends React.PureComponent<DatasetSettingsProps, State> {
+          <br />
+          This will overwrite the current default view configuration.
+          <br />
+          This includes the current camera position, zoom level, interpolation setting and layer
```
So: Here are the results of my investigation: All settings you mentioned are sent to the backend and upon loading the dataset in view mode all settings are returned from the backend except for the position, zoon, and maybe rotation. But I think the default position and zoom and rotation are sent by another request to the frontend. Although the position does not seem to be applied upon loading the dataset (without the position already being defined by the query params)

Note that I did not test the `segmentationPatternOpacity` and the `rotation property.

This makes the new button more potent than the settings UI as the settings UI does not allow to configure the following settings:
```
fourBit
renderMissingDataBlack
loadingStrategy
blendMode
```

> I'm wondering whether we should make it more explicit what exactly would be saved.

For this I can add a table to the modal that tasks for confirmation. The table would list the properties saved. 

To agree upon what settings should be saved we can talk in person :)

MichaelBuessemeyer commented:
> The layer configuration in the dataset view settings doesn't currently allow to change: brightness, contrast, min, max
I think these should be added to the schema so they can be changed (and documentation in the table at the right should be added, too).

I am pretty sure that brightness and contrast were replaced with `min, max, intensityRange` and thus they should not be configurable. But you are right with `min` and `max`. They are already configurable via the json layer config in the UI and are now also added to the table.

As discussed I reworked the UI and made the `intensityRange` setting optional / required for color layers.
As this change required some changes in multiple files, it would be great if you could look over them to avoid accidental bugs :pray:

daniel-wer commented on frontend/javascripts/types/schemas/dataset_view_configuration_defaults.ts:
```diff
@@ -62,6 +63,11 @@ export const enforceValidatedDatasetViewConfiguration = (
+        if (layer.category === "segmentation") {
+          delete existingLayerConfig.intensityRange;
+        } else if (existingLayerConfig.intensityRange == null) {
+          existingLayerConfig.intensityRange = getDefaultIntensityRangeOfLayer(dataset, layer.name);
```
Why is this always added for color layers? In my tests, I saved a layer configuration of 

```
{
  "color": {},
  "segmentation": {} 
}
```

and when I reloaded the page, the layer configuration was:

```
{
  "color": {
    "intensityRange": [
      0,
      255
    ]
  },
  "segmentation": {}
}
```

This was unexpected.

daniel-wer reviewed (commented):
Thank you for the newest improvements. I have a couple of small remarks, but afterwards this PR should be ready to be merged :+1:

MichaelBuessemeyer reviewed (commented):
Thanks for the review daniel. I hope everything should be fine now.

daniel-wer commented on frontend/javascripts/oxalis/model/sagas/clip_histogram_saga.ts:
```diff
@@ -11,11 +11,7 @@ import { api } from "oxalis/singletons";
-  } else {
-    Store.dispatch(updateLayerSettingAction(layerName, "intensityRange", [firstVal, secVal]));
-  }
+  Store.dispatch(updateLayerSettingAction(layerName, "intensityRange", [firstVal, secVal]));
```
Sounds good, thank you for testing!

daniel-wer reviewed (approved):
Nice, LGTM :+1:

MichaelBuessemeyer commented:
Sadly, (but obviously) this pr did not work with #7188 out of the box. I had to add a missing form item that enables the user to manipulate and save the order of the color layers. 
As an approach, I used the same dnd lib as used in #7188 to create a list of sortable color layer names via dnd.
Sadly this also implies that I had to add another settings item to the UI. This means by the current layout that a new row only for a single setting needed to be added.

The current version looks like this: 
![image](https://github.com/scalableminds/webknossos/assets/39529669/c20c2eaa-f398-4fea-8f76-c91b5f234fd1)
(Below the part of the image is a divider and then the json settings input for each layer. I excluded this in the screenshot.)
As the list takes up a lot of space on the screen, I wrapped it in a collapsable and hide it as default. The collapsible is also only extendable when having more than one color layer.

 @philippotto @daniel-wer Do you think this UI is okayish enough? / Do you have suggestions?

philippotto commented on frontend/javascripts/messages.tsx:
```diff
@@ -79,6 +80,8 @@ export const settingsTooltips: Partial<Record<keyof RecommendedConfiguration, st
   renderWatermark: "Show a WEBKNOSSOS logo in the lower-left corner of each screenshot.",
   antialiasRendering: "Antialias rendering (can impact performance)",
+  colorLayerOrder:
+    "Set the order in which color layers are rendered. This setting will only affect the cover blend mode.",
```
```suggestion
    "Set the order in which color layers are rendered. This setting is only relevant if the cover blend mode is active.",
```

is a bit clearer in my opinion.

philippotto commented on frontend/javascripts/dashboard/dataset/color_layer_ordering_component.tsx:
```diff
@@ -0,0 +1,74 @@
+
+  const isSettingEnabled = colorLayerNames && colorLayerNames.length > 1;
+  const collapsibleDisabledExplanation =
+    "This setting only makes sense if there are multiple color layers.";
```
```suggestion
    "The order of layers can only be configured when the dataset has multiple color layers.";
```

philippotto commented:
Code looks good, but I couldn't change the layer order for this DS: https://updatedefaultlayerconfiginviewmode.webknossos.xyz/datasets/sample_organization/idr0101A-4d/edit
even though it has multiple color layers?

MichaelBuessemeyer commented on frontend/javascripts/types/schemas/dataset_view_configuration_defaults.ts:
```diff
@@ -71,6 +98,7 @@ export const enforceValidatedDatasetViewConfiguration = (
           : _.pickBy(layerConfigDefault, (value: any) => value !== null);
       }
     });
+    ensureDatasetSettingsHasLayerOrder(datasetViewConfiguration, dataset);
```
> Code looks good, but I couldn't change the layer order for this DS: https://updatedefaultlayerconfiginviewmode.webknossos.xyz/datasets/sample_organization/idr0101A-4d/edit
even though it has multiple color layers?

Thanks for finding this bug :mag:.

This line should fix the bug (according to my testing). 
The problem was I added the new property `colorLayerOrder` (see https://github.com/scalableminds/webknossos/pull/7188) and that it is expected to be always present in a validated `DatasetConfiguration`. But the backend does not send this property if it wasn't set by the frontend previously. 
Previous to my changes, the property was only enforced during model initialization (see changes in `model initialization`). But the enforcing of the property with correct values must be done every time when working with a `DatasetConfiguration`. Thus it made more sense to me, to move the enforcing code to the `enforceValidatedDatasetViewConfiguration` function. 
This covers the bug, as this function is explicitly called in `dataset_settings_view.tsx` after it is fetched from the backend. 
Additionally, the `enforceValidatedDatasetViewConfiguration` is also enforced for the model initialization as it is part of the `getDatasetViewConfiguration` API backend call in `admin_rest_api.ts` which is called during the model initialization to fetch the `DatasetConfig` from the backend.

Could you please retest the changes (once the CI is done)?

philippotto replied:
Hm, it still doesn't work for me [here](https://updatedefaultlayerconfiginviewmode.webknossos.xyz/datasets/sample_organization/idr0101A-4d/edit)? :thinking: Does it work for you?

philippotto replied:
Great, now it works :)
